Add a contribution ladder to gittuf (#11)
gittuf/community: CONTRIBUTOR_LADDER.md

# gittuf Contributor ladder

gittuf's contributor ladder is loosely inspired by the OpenSSF [Scorecard
Contributor Ladder]. This document details how community members may advance to
more-involved roles within the gittuf project.

- [Roles](#roles)
    - [Community members](#community-members)
    - [Maintainers](#maintainers)
- [Inactive members](#inactive-members)

## Roles

### Community members

Community members engage with gittuf, contributing their time and energy in
discussions or just generally helping out.

#### Pre-requisites

- Must follow the [OpenSSF Code of Conduct]
- Must follow the [Contribution Guide]

#### Responsibilities

- Keep it up!

### Maintainers

Maintainers are responsible for the project's overall health. They are the only
ones who can approve and merge code contributions.

**Defined by:** Listed in the `MAINTAINERS.txt` file in the project repository

#### Pre-requisites

> - Reviewed at least 5 substantial PRs to the codebase
> - Reviewed or got at least 10 PRs merged to the codebase

Reviewer: Do we want literal limits or should this be more like my suggested text:

```suggestion
- Provided substantial improvements to the codebase (e.g., adding 5 substantial PRs to the codebase)
- Helped to review other submissions to the codebase (e.g., reviewed 10 PRs that were merged to the codebase)
```

Reviewer: I also think both of the above lines in your PR should not start with Reviewed.  Do you mean authored, etc.?

Author: Agreed on both points, I'll update the PR.

# gittuf Contributor ladder

gittuf's contributor ladder is loosely inspired by the OpenSSF [Scorecard
Contributor Ladder]. This document details how community members may advance to
more-involved roles within the gittuf project.

- [Roles](#roles)
    - [Community members](#community-members)
    - [Maintainers](#maintainers)
- [Inactive members](#inactive-members)

## Roles

### Community members

Community members engage with gittuf, contributing their time and energy in
discussions or just generally helping out.

#### Pre-requisites

- Must follow the [OpenSSF Code of Conduct]
- Must follow the [Contribution Guide]

#### Responsibilities

- Keep it up!

### Maintainers

Maintainers are responsible for the project's overall health. They are the only
ones who can approve and merge code contributions.

> **Defined by:** "Maintain" permissions in the project

Reviewer 2: ```suggestion
**Defined by:** Listed in the `MAINTAINERS.txt` in the project repository
```

#### Pre-requisites

> - Reviewed at least 10 substantial PRs to the codebase

Reviewer 2: ```suggestion
- Reviewed at least 5 substantial PRs to the codebase
```

> - Reviewed or got at least 30 PRs merged to the codebase

Reviewer 2: 30 seems too high given our project velocity.

Author: Would 10 or 15 work here?

Author: I've updated this to be 10 PRs for the latter point.

> - Helped to review other submissions to the codebase (e.g., reviewed 10 PRs that

Reviewer 2: I'd reuse "substantial" here too to communicate that dependabot PRs (for example) don't really count. The word is underdefined but I think for now we can go with "you know it when you see it".

#### Responsibilities

- Demonstrate sound technical judgment
- Maintain project quality control via code reviews

>     - Focus on holistic acceptance of contribution

Reviewer 2: What does this mean?

Author: Took this out, it is unclear.

- Be responsive to review requests
- Mentor contributors

> - Approve and merge code contributions as appropriate

Reviewer 2: Isn't this covered in the code reviews bit above?

- Participate in gittuf community meetings, if possible

> - Facilitating gittuf-specific community meetings, if possible and comfortable

Reviewer 2: ```suggestion
- Facilitate gittuf community meetings, if possible and comfortable
```

> - Facilitate gittuf community meetings, if possible and comfortable

Reviewer: ```suggestion
- Facilitate gittuf community meetings, as is needed
```

#### Privileges

> - Maintainer status may be a precondition to accepting especially large code

Reviewer 2: I'd take this out as a  whole, I'm not sure we want to grant this privilege (even though I'm aware I've been guilty of this on occasion).

  contributions

#### Promotion process

> - Sponsored by a maintainer

Reviewer 2: ```suggestion

- Sponsored by a maintainer
```

    - With no objections from other maintainers

>     - Done through PR to update the MAINTAINERS.md file

Reviewer 2: ```suggestion
    - Done through PR to update the `MAINTAINERS.txt` file
```

> - May self-nominate or be nominated by a maintainer

Reviewer 2: How is nomination different from sponsorship?

Author: I figure we can just combine these into being nominated.

Author: Cleared this up.

    - In case of self-nomination, sponsor must comment approval on the PR

## Inactive members

> A core principle in maintaining a healthy community is encouraging active

Reviewer 2: ```suggestion

A core principle in maintaining a healthy community is encouraging active
```

participation. It is inevitable that a contributor's focus will change over time
and there is no expectation they'll actively contribute forever.

Any contributor at any level described above may write an issue (or PR, if

> MAINTAINERS.md changes are necessary) asking to step down to a lighter-weight

Reviewer 2: ```suggestion
`MAINTAINERS.txt` changes are necessary) asking to step down to a lighter-weight
```

Reviewer 2: Also I'd note that we will maintain an emeritus list for maintainers!

tier or to depart the project entirely. Such requests will hopefully come after
thoughtful conversations with the rest of the team and with sufficient
forewarning for the others to prepare. However, sometimes "life happens".
Therefore, the change in responsibilities will be understood to take immediate
effect, regardless of whether the issue/PR has been acknowledged or merged.

However, should a Maintainer be deemed inactive for a significant period, any
Community Member or above may write an issue/PR requesting their removal from
the ranks (and `@mentioning` the inactive contributor in the hopes of drawing
their attention). The request must receive support (in comments) from a majority
of Maintainers to proceed.

> [OpenSSF Code of Conduct]: https://openssf.org/community/code-of-conduct/

Reviewer 2: ```suggestion
[OpenSSF Code of Conduct]: /CODE-OF-CONDUCT.md
```
